kernel/core/datamanagement/memory.hpp

```diff
@@ -218,6 +219,432 @@ class Memory : public Sorter
+            FactorNode(const std::vector<size_t> s) : subset(s) {}
+
+            /////////////////////////////////////////////////
+            /// \brief This function adds a chid element to the FactorNode
+            ///
+            /// \param child FactorNode*
+            /// \return void
+            ///
+            /////////////////////////////////////////////////
+            void addChild(FactorNode* child)
+            {
+                children.push_back(child);
+            }
+
+            /////////////////////////////////////////////////
+            /// \brief Destructor of FactorNode
+            ///
+            ///
+            /////////////////////////////////////////////////
+            ~FactorNode()
+            {
+                for (auto child : children)
+                    delete child;
+
+                // ne wir kommen ja vom parent ?
+                //if(parent != nullptr)
+                //    delete parent;
+            }
+        };
+
+        class AnovaCalculationStructure
+        {
+        private:
+            FactorNode* root;
+            Memory* mem;
+            double significance = 0;
+            std::vector<std::vector<std::string>> factors;
+
+            mu::value_type overallMean;
+            mu::value_type overallNum;
+            mu::value_type SS_Within;
+            double dof_within = 0;
+            size_t max_depth = 0;
+
+            /////////////////////////////////////////////////
+            /// \brief This helper function does build the Tree up to a given level n
+            ///
+            /// \param node FactorNode*
+            /// \param start int
+            /// \param n int
+            /// \param currentSet std::vector<size_t>&
+            /// \return void
+            ///
+            /////////////////////////////////////////////////
+            void buildTreeHelper(FactorNode* node, int start, int n, std::vector<size_t>& currentSet)
+            {
+                if (start > n) return; // Base case: no more elements to add
+
+                //depth is given by size of subset
+                if(max_depth < currentSet.size()+1)
+                    max_depth = currentSet.size()+1;
+
+                for (size_t i = start; i <= n; ++i)
+                {
+                    std::vector<size_t> newSubset = currentSet;
+                    newSubset.push_back(i);
+                    FactorNode* child = new FactorNode(newSubset);
+
+                    // we are at level > 2
+                    if(currentSet.size() > 0)
+                        child->parent = node;
+
+                    std::stringstream ss;
+                    for(auto i : newSubset)
+                        ss << i << ",";
+                    std::string test = ss.str();
+
+                    //calculate SS for new child
+                    calculateMean(child, i);
+                    calculateDof(child, factors[i].size());

```

Comment:
The original idea was that the per-child calculations are done on child level (and those apply them recursively), i.e. after creating the tree, call `root->calculateMean(); root->calculateDoF();`

If you follow this approach, it should free you from having additional navigation structures as you always apply calculations recursively (i.e. `calculateLevel()` will be part of a node and therefore already knows its level). It may, however, be possible that there are cross-branch dependencies in the calculations (which is sub-optimal). Then you cannot follow this approch, of course. Still, I'd recommend to move all per-node calculations to their node.

Reply:
Good Point!
The `calculateMean`() does use the shared pointer to the memory element containing all data. however that is just a pointer I could forward this also to the Node element.
calculateDoF can also be changed to the node sure!
actually the same is true for `calculateSS()`.

However,` calculateSSInteraction(FactorNode* node)`  does also need Uncle, Great Uncle Node, ...
so here the whole calculations of each previous level have to be already completed. This is why I needed to go with this extra travel trhough the tree with `calculateLevel()`.

So my though was to have all calculations in the `AnovaCalculationStructure` rather then split it up half half. 

Should I put everything that is possible into the Node class?

Reply:
Probably no, but isn't it possible to delegate the calculations, which need now "uncle" branches to their parents? In the end, you'll reach `root`, which is the common parent and which doesn't have an uncle node. Sooo... at least it looks like it should be possible, no?

It's a design decision to have all together or all split up. I'd vote for divide n' conquer.

Reply:
Oh sure the traversal is possible! The thing is, that not all great uncle nodes do already exist at the point where the tree structure is created. 
Hence, the calculation of SSInteractions can only be applied in a second iteration through the tree.
In the second run it is most straight forward to trigger the calculations level after level to ensure all SSinteraction values needed at the current Node are already calculated. Here we can then get all needed Nodes for the calculation starting at the root node.

The second approach would be travel through the tree "normally" and get all Nodes needed from the current node. Here we would travel back up till the root element, while also going down at each element to find the needed Nodes in every sub part of the tree. Here the traversal to reach all needed nodes would take exactly the same amount of steps, but would be needed to split up in two helper functions.  Then we will get to some Nodes, where SSInteraction is not calculated yet, and this calculation first has to be triggered again. This would yield in a lot more "recursion levels", and probably more complex code.

However, going level by level means we will have redundant travels for each level! 
I would argue it is not worth to increase complexity of the SSinteraction calculation at this point, since in reality the AnovaTreeStructure will never get that big that it indeed is a big impact when calculation is applied level after level.

Reply:
> It's a design decision to have all together or all split up. I'd vote for divide n' conquer.

Hmmm. When adding calculations to the nodes, now we have to pass all shared parameters of all nodes to every single node for the calculations. OverallMean, OverallNum, the memory object holding all single elements.
Actually there is only the calculate DoF function, which does not need to access a shared element of the whole tree structure. 
For me, it makes the most 'sense' to only have things in the Node that clearly concern that one Node alone.

Reply:
Understood, what is about `node.calculteMean(const Memory&)` or something similar? But yeah, do not over-engineer it.

Reply:
I did restructure  `calculateSS`, `calculateMean` and `calculateDof` into the Node and call it directly when building the Tree.
However, I would argue to keep `calculateSSInteraction` and `SSWithin` in the second calculation iteration through the tree since otherwise it would need a bit more complex recursive structure for tree-traversal.

Also the ANOVA documentation articles need to be updated right?

Reply:
Accepted.

Articles, Language flles, find.ndb entries, yes. In the latter it's probably only a signature update, nothing new to be added